pageserver/src/tenant/pricing.rs

```diff
@@ -0,0 +1,458 @@
+use std::collections::{HashMap, HashSet};
+use std::sync::Arc;
+
+use anyhow::Context;
+use tokio::sync::Semaphore;
+
+use super::Tenant;
+use utils::id::TimelineId;
+use utils::lsn::Lsn;
+
+use tracing::*;
+
+/// Inputs to the actual pricing model
+///
+/// Implements [`serde::Serialize`] but is not meant to be part of the public API.
+#[serde_with::serde_as]
+#[derive(Debug, serde::Serialize)]
+pub struct ModelInputs {
+    updates: Vec<Update>,
+    retention_period: u64,
+    #[serde_as(as = "HashMap<serde_with::DisplayFromStr, _>")]
+    timeline_inputs: HashMap<TimelineId, TimelineInputs>,
+}
+
+/// Collect all relevant LSNs to the inputs, even though they will only be helpful in the
+/// serialized form.
```

Comment:
I don't understand this comment. How are they only helpful in the serialized form?

Reply:
Yes, I was hinting towards manual inspection happening later on the returned document from the http api, but this comment is not great either, need to think about this...

Reply:
Elaborated on the manual inspection hint.